### PR TITLE
Cleanup bart caching logic

### PR DESCRIPTION
**Pull request by sshleifer**
Previously, we had a helper function that checked 4 possible cases to determine whether we should:

(a) combine a cached attention mask with a new one.
(b) just use the cached one
(c) just use the new/passed one

This consolidates that logic into 3 branches and deletes the helper func, which was only called once.

### CODE REVIEW

File: src/transformers/modeling_bart.py [hunk 1]
```diff
@@ -694,37 +693,14 @@ def _use_saved_state(self, k, v, saved_state, key_padding_mask, static_kv, bsz):
...
-    @staticmethod
-    def _cat_prev_key_padding_mask(
-        key_padding_mask: Optional[Tensor],
-        prev_key_padding_mask: Optional[Tensor],
-        batch_size: int,
-        src_len: int,
-        static_kv: bool,
-    ) -> Optional[Tensor]:
-        # saved key padding masks have shape (bsz, seq_len)
         if prev_key_padding_mask is not None:
```
Comment (sshleifer) [inline src/transformers/modeling_bart.py:696]: cases a,b are unchanged.

File: src/transformers/modeling_bart.py [hunk 2]
```diff
@@ -694,37 +693,14 @@ def _use_saved_state(self, k, v, saved_state, key_padding_mask, static_kv, bsz):
...
             if static_kv:
                 new_key_padding_mask = prev_key_padding_mask
             else:
                 new_key_padding_mask = torch.cat([prev_key_padding_mask, key_padding_mask], dim=1)
-
-        elif key_padding_mask is not None:
-            filler = torch.zeros(
-                batch_size,
-                src_len - key_padding_mask.size(1),
```
Comment (sshleifer) [inline src/transformers/modeling_bart.py:720]: in every case, `filler.shape[1] == 0` (filler is an empty tensor) so this `torch.cat` is unneeded.

File: src/transformers/modeling_bart.py [hunk 3]
```diff
@@ -648,10 +648,9 @@ def forward(
         # get here for encoder decoder cause of static_kv
         if layer_state is not None:  # reuse k,v and encoder_padding_mask
             saved_state = layer_state.get(self.cache_key, {})
-            if "prev_key" in saved_state:
+            if "prev_key" in saved_state and static_kv:
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bart.py:651]: what does `static_kv` mean excatly ?

Comment (patrickvonplaten) [reply src/transformers/modeling_bart.py:651]: Like "no need" to recalculate key-value projections?

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/modeling_bart.py (hunk 3 above)
Comment (sshleifer) [reply src/transformers/modeling_bart.py:651]: exactly. Notation copied from fairseq.
